CRM/Financial/Form/FrontEndPaymentFormTrait.php

```diff
@@ -0,0 +1,69 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | CiviCRM version 5                                                  |
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC (c) 2004-2019                                |
+ +--------------------------------------------------------------------+
+ | This file is a part of CiviCRM.                                    |
+ |                                                                    |
+ | CiviCRM is free software; you can copy, modify, and distribute it  |
+ | under the terms of the GNU Affero General Public License           |
+ | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
+ |                                                                    |
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC (c) 2004-2019
+ */
+
+/**
+ * This class holds functionality shared between various front end forms.
+ */
+trait CRM_Financial_Form_FrontEndPaymentFormTrait {
+
+  /**
+   * Alter line items for template.
+   *
+   * This is an early cut of what will ideally eventually be a hooklike call to the
+   * CRM_Invoicing_Utils class with a potential end goal of moving this handling to an extension.
+   *
+   * @param $tplLineItems
+   *
+   * @return array
+   */
+  protected function alterLineItemsForTemplate($tplLineItems) {
+    $getTaxDetails = FALSE;
+    foreach ($tplLineItems as $key => $value) {
+      foreach ($value as $k => $v) {
+        if (isset($v['tax_rate'])) {
+          if ($v['tax_rate'] != '') {
```

Comment:
Can we collapse this if down to `if(!empty($v['tax_rate'])` as that's exactly what the 2 if statements here are doing

Reply:
@mattwire what if the tax rate is 0% hypothetically like what if some line items have tax applied but not others?

Reply:
@mattwire Yes we can - I'm happy to add it in here but there is other cleanup to do in this function so I deliberately left the cleanup out of the first round to keep it simple

Reply:
@seamuslee001 that's fine - if ANY rows have tax then the getTaxDetails = TRUE is assigned to the template 

A good follow up question is 'what does getTaxDetails mean'

I see it used in a few places - where I would probably say it's the wrong way to assign 

```
        {if $getTaxDetails}
          <th class="right">{ts}Subtotal{/ts}</th>
          <th class="right">{ts}Tax Rate{/ts}</th>
          <th class="right">{ts}Tax Amount{/ts}</th>
          <th class="right">{ts}Total Amount{/ts}</th>
        {/if}
```

And which probably rely on the variable being assigned somewhere completely different - so it's likely absolutely nothing is achieved.....

Reply:
@mattwire Actually in the cleanup round changing from 2 lines to the same 2 conditions on one line makes more sense as adding assumptions that the template doesn't want to know about zero is probably a bigger leap - but I'd rather cleanup as PR 2